source/_ecosystem/oauth2.markdown

```diff
@@ -0,0 +1,202 @@
+release_date: 2017-1-22 12:00:00 -0600
+sidebar: true
+comments: true
+sharing: true
+footer: true
+---
+
+Once you have [NGINX](https://home-assistant.io/ecosystem/nginx/) set up with Home Assistant, your own domain, and an SSL Certificate, you can set up authentication to require sign in with an oauth2 authentication provider suchas Google, Github, Linkedin, and others using [bitly\oauth2_proxy](https://github.com/bitly/oauth2_proxy).
+
+This also allows 2-factor authentication when using an oauth2 providers that support it.
+
+These instructions will walk you through modifying your nginx configuration to support one of these providers by setting up your own oauth2_proxy and using the auth_request directive.
+
+### {% linkable_title 1. Pick your oauth provider and acquire the authentication keys. %}
+
+This is taken directly from the oauth2_proxy [Github Page](https://github.com/bitly/oauth2_proxy). For other providers, visit that page as it has more detailed instructions.
+
+For Google, the registration steps are:
+
+1. Create a new project: https://console.developers.google.com/project
+2. Choose the new project from the top right project dropdown (only if another project is selected)
+3. In the project Dashboard center pane, choose **"Enable and manage APIs"**
+4. In the left Nav pane, choose **"Credentials"**
+5. In the center pane, choose **"OAuth consent screen"** tab. Fill in **"Product name shown to users"** and hit save.
+6. In the center pane, choose **"Credentials"** tab.
+   * Open the **"New credentials"** drop down
+   * Choose **"OAuth client ID"**
+   * Choose **"Web application"**
+   * Application name is freeform, choose something appropriate
+   * Authorized JavaScript origins is your domain ex: `https://internal.yourcompany.com`
+   * Authorized redirect URIs is the location of oath2/callback ex: `https://internal.yourcompany.com/oauth2/callback`
+   * Choose **"Create"**
+4. Take note of the **Client ID** and **Client Secret**
+
+### {% linkable_title 2. Install oauth2_proxy %}
+
+I am going to take the cheap route and send you directly to the [Github Page](https://github.com/bitly/oauth2_proxy) for oauth2_proxy for general installation of the product.
+
+If you are using docker, you can use this command to run it and reference the configuration files we create in the next step:
+
+```
+ docker run -d --name="oauth2_proxy" --net="bridge" -e -p 8095:8095/tcp -v "/path/to/appdata/oauth2_proxy":"/config":rw bitly/oauth2_proxy -config="/config/oauth2_proxy.cfg"
+```
+
+### {% linkable_title 3. Configure oauth2_proxy %}
+
+We are going to create 2 files, "oauth2_proxy.cfg" and "auth_emails", which will contain a list of email addresses that will be allowed in.
+
+In oauth2_proxy.cfg, the biggest non-standard thing we are going to do is to proxy it to file:///dev/null so that it doesn't attempt to return anything other than the 200 OK response to nginx because oauth2_proxy is only being used to auth, not to proxy.
```

Comment:
For file names and paths we usually like to make it look like code. You can do this by surrounding a string with backticks (`) i.e. file:///dev/null should instead be `file:///dev/null`. May want to do the same for `oauth2_proxy.cfg` etc.